yt/chyt/server/query_analyzer.cpp

```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
+        // and even monotonic functions, I do not think there is much use.
+        auto columnName = orderByElement->children.front()->getColumnName();
+
+        // Correct index range and sort order presence is guaranteed by the key column count above.
+        const auto& schemaColumn = schema->Columns()[columnIndex];
+        YT_VERIFY(schemaColumn.SortOrder());
+
+        // Columns from the ORDER BY clause must form a prefix of the table's primary key.
+        if (columnName != schemaColumn.Name()) {
+            return EReadInOrderMode::None;
+        }
+
+        // All requested directions must align and form either a forward or backward read of the underlying table.
+        auto columnDirection = GetReadInOrderMode(*schemaColumn.SortOrder(), orderByElement->direction);
+        if (commonDirection == EReadInOrderMode::None) {
+            commonDirection = columnDirection;
+        } else if (commonDirection != columnDirection) {
+            return EReadInOrderMode::None;
+        }
+
+        // See float-related comments below.
+        if (IsFloatingPointType(schemaColumn.CastToV1Type()) && !orderByElement->nulls_direction_was_explicitly_specified) {
+            floatKeyColumnsWithoutExplicitNullsDirection.push_back(schemaColumn.Name());
+        }
+    }
+
+    // See the last comment in this function.
+    if (!floatKeyColumnsWithoutExplicitNullsDirection.empty()) {
+        THROW_ERROR_EXCEPTION(
+            "Nulls direction must be specified explicitly for key columns of floating-point "
+            "types when using optimize_read_in_order is enabled; see docs for more details")
+            << TErrorAttribute("problematic_columns", floatKeyColumnsWithoutExplicitNullsDirection);
+    }
+
+    // In YT, NULL values compare less than any other values. In ClickHouse, NULL value placement
+    // depends solely on the placement requested in the query, which defaults to NULLS LAST.
+    // Since we combine both YT and CH sorting in this mode, there is only one acceptable direction
+    // depending on the requested sort order. In cases when the unsupported option turns out to be
+    // the default (for ascending sorts), we tweak the query to avoid additional user turmoil.
+    if (commonDirection != EReadInOrderMode::None) {
+        for (const auto& orderByElementAst : selectQuery->orderBy()->children) {
+            auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+            // The value of nulls_direction is equal to direction for NULLS LAST and to the opposite
+            // of direction for NULLS FIRST.
+            // For ASC (d = 1), we need NULLS FIRST (nd = -d = -1).
+            // For DSC (d = -1), we need NULLS LAST (nd =  d = -1).
+            if (orderByElement->nulls_direction != -1) {
+                THROW_ERROR_EXCEPTION_IF(
+                    orderByElement->nulls_direction_was_explicitly_specified,
```

Comment:
We have discussed that there should be no errors here - just disable the optimization if the order by clause is incompatible.

Reply:
Removed this, see other comments.